pkg/azdiskschedulerextender/azdiskschedulerextender.go

```diff
@@ -0,0 +1,199 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog/v2"
+	schedulerapi "k8s.io/kube-scheduler/extender/v1"
+
+	clientSet "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned/typed/azuredisk/v1alpha1"
+)
+
+func init() {
+
+	var err error
+	azVolumeAttachmentExtensionClient, azDriverNodeExtensionClient, err = getKubernetesClient()
+
+	if err != nil {
+		klog.Fatalf("Failed to create kubernetes clinetset %s ...", err)
+		os.Exit(1)
+	}
+}
+
+func filter(schedulerExtenderArgs schedulerapi.ExtenderArgs) (*schedulerapi.ExtenderFilterResult, error) {
+	var (
+		filteredNodes     []v1.Node
+		filteredNodeNames []string
+		failedNodes       map[string]string
+	)
+
+	// all available cluster nodes
+	allNodes := schedulerExtenderArgs.Nodes.Items
+	// all nodes that have azDriverNode running
+	azDriverNodes, err := azDriverNodeExtensionClient.List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get the list of azDriverNodes: %v", err)
+	}
+
+	// map node name to azDriverNode state
+	nodeNameToStatusMap := make(map[string]string)
+	for _, azDriverNode := range azDriverNodes.Items {
+		nodeNameToStatusMap[azDriverNode.Spec.NodeName] = azDriverNode.Status.State
+	}
+
+	// Filter the nodes based on AzDiverNode status
+	failedNodes = make(map[string]string)
+	for _, node := range allNodes {
+		state, ok := nodeNameToStatusMap[node.Name]
+		if ok && state == "Ready" {
+			filteredNodes = append(filteredNodes, node)
+			filteredNodeNames = append(filteredNodeNames, node.Name)
+		} else {
+			failedNodes[node.Name] = fmt.Sprintf("AzDriverNodes for %s is not ready.", node.Name)
+		}
+		klog.V(2).Infof("handleFilterRequest: %v %+v", node.Name, node.Status.Addresses)
+	}
+
+	return formatFilterResult(filteredNodes, filteredNodeNames, failedNodes, ""), nil
+}
+
+func prioritize(schedulerExtenderArgs schedulerapi.ExtenderArgs) (priorityList schedulerapi.HostPriorityList, err error) {
+
+	availableNodes := schedulerExtenderArgs.Nodes.Items
+	requestedVolumes := schedulerExtenderArgs.Pod.Spec.Volumes
+
+	//TODO add RWM volume case here
+	// if no volumes are requested, return assigning 0 score to all nodes
+	if requestedVolumes == nil {
+		for _, node := range availableNodes {
+			hostPriority := schedulerapi.HostPriority{Host: node.Name, Score: 0}
+			priorityList = append(priorityList, hostPriority)
+		}
+	} else {
+		volumesPodNeeds := make(map[string]bool)
```

Comment:
Yes, you are right. Was reading about this in the context of channels, where `struct{}{}` was used to communicate back to main goroutine instead of `bool`.